test/test_sparse_csr.py

```diff
@@ -2209,8 +2241,91 @@ def test_dense_to_from_sparse_compressed(self, device, layout):
+            _test_matrix(pt_matrix, dense, layout, blocksize)
             self.assertEqual(dense, pt_matrix.to_dense())
 
+        if layout is not torch.sparse_bsr:
+            # TODO: Remove this once support has been enabled
+            return
+
+        # Test batch shapes (ND inputs)
+
+        # Case 1: Same sparsity pattern across matrices
+        for shape, blocksize, batch_shape in itertools.product(shapes, blocksizes, batch_sizes):
+            full_shape = batch_shape + shape
+            batch_len = functools.reduce(lambda x, y: x * y, batch_shape, 1)
+            dense = make_tensor(full_shape, dtype=torch.float, device=device)
+            # select the first batch to create the mask
+            mask = dense[tuple(np.unravel_index(0, batch_shape))].relu().bool()
+            dense = dense * mask
+            pt_tensor = self._convert_to_layout(dense, layout, blocksize=blocksize)
+            for i in range(batch_len):
+                batch_idx = tuple(np.unravel_index(i, batch_shape))
+                _test_matrix(pt_tensor[batch_idx], dense[batch_idx], layout, blocksize)
+            # todo: check whole conversion once to_dense impl for n-d batched-bsr
+            # take 3d slices of dense/sparse to convert/compare for now
+            if dense.dim() > 3:
+                part_dim = dense.dim() - 3
+                part_shape = batch_shape[:part_dim]
+                len_partition = functools.reduce(lambda x, y: x * y, part_shape, 1)
+                for i in range(len_partition):
+                    part_idx = tuple(np.unravel_index(i, part_shape))
+                    self.assertEqual(dense[part_idx], pt_tensor[part_idx].to_dense())
+            else:
+                self.assertEqual(dense, pt_tensor.to_dense())
+
+        # Verify exception when given 0 sized batch
+        for shape, blocksize in itertools.product(shapes, blocksizes):
+            dense = make_tensor((0,) + shape, dtype=torch.float, device=device)
+            # TODO: Support zero sized batch dimensions
+            with self.assertRaisesRegex(RuntimeError, "to_sparse_bsr: Expected product of batch dimensions to be non-zero."):
+                self._convert_to_layout(dense, layout, blocksize=blocksize)
+
+        # TODO: Case 2: Different sparsity pattern across matrices, but same number of zeros
+        # NOTE: For blocksparse formats this applies at a per-block level,
+        dense = make_tensor((2, 4, 4), dtype=torch.float, device=device)
+        blocksize = (2, 2)
+        mask = torch.tensor([
+            [[True, True], [False, True]],
+            [[True, False], [True, True]]],
+            device=device).view((2, 2, 2, 1, 1))
+        mask = mask.expand((2, 2, 2, 2, 2))
+        mask = mask.transpose(2, 3)
+        mask = mask.reshape_as(dense)
+        dense = dense * mask
+        if layout == torch.sparse_bsr:
+            # this is not an error as long as the nse is equal for bsr
```

Comment:
I want to highlight this for reviewers. 

The original 3-D extension by @cpuhrsch required that all batches have the same sparsity pattern, that is Both the number **and** location of specified blocks had to be the same. This is at odds with the blog post by @pearu on BSR invariants which states a more relaxed requirement.

> ...different batches of BSR tensors may have different sparsity patterns as long as the number of specified elements is the same for all batches.

I have implemented this with the more permissive requirement from the blog post, but it is fairly simple to reintroduce the strict check if needed.

Reply:
@amjames - the additional constraints were only meant as a stepping stone to the more generic version that you now implemented :)